lib/services/elasticsearch.js

```diff
@@ -488,70 +509,44 @@ class ElasticSearch extends Service {
-      return Bluebird.reject(new BadRequestError('Query cannot be empty'));
+    if (!request.input.args.includeTrash) {
+      addActiveFilter(esRequest);
     }
 
-    return getAllIdsFromQuery(this.client, addActiveFilter(esRequestSearch))
-      .then(ids => {
-        ids.forEach(id => {
-          esRequestBulk.body.push({update: {_index: esRequestBulk.index, _type: esRequestBulk.type, _id: id}});
-          esRequestBulk.body.push({doc: {_kuzzle_info: { active: false, deletedAt: Date.now(), updater: request.context.user._id ? String(request.context.user._id) : null}}});
-        });
-
-        if (esRequestBulk.body.length === 0) {
-          return Bluebird.resolve({ids: []});
-        }
-
-        return this.client.bulk(esRequestBulk)
-          .then(() => this.refreshIndexIfNeeded(esRequestBulk, {ids}))
-          .catch(error => Bluebird.reject(this.esWrapper.formatESError(error)));
-      });
-  }
-
-  /**
-   * Delete all document that match the given filter from the trash
-   * @param {Request} request
-   * @returns {Promise} resolve the list of deleted ids
-   */
-  deleteByQueryFromTrash(request) {
-    const
-      esRequestSearch = initESRequest(request, ['from', 'size', 'scroll']),
-      esRequestBulk = initESRequest(request, ['refresh']);
-
-    assertWellFormedRefresh(esRequestBulk);
-
-    esRequestSearch.body = request.input.body;
-    esRequestSearch.scroll = '30s';
-
-    if (esRequestSearch.body.query === null) {
-      return Bluebird.reject(new BadRequestError('null is not a valid document ID'));
+    if (request.input.args.bypassTrash) {
+      // *real* deletion
+      esMethod = req => this.client.deleteByQuery(req);
+      responseMember = 'deleted';
+    }
+    else {
+      esRequest.body.script = {
+        lang: 'painless',
+        stored: '_kuzzle_info_delete',
+        params: {now: Date.now()}
+      };
+      esRequest.conflicts = request.input.args.conflicts || 'abort';
     }
 
-    return getAllIdsFromQuery(this.client, esRequestSearch)
-      .then(ids => {
-        return new Bluebird((resolve, reject) => {
-          esRequestBulk.body = ids.map(id => ({delete: {_index: esRequestBulk.index, _type: esRequestBulk.type, _id: id}}));
-
-          if (esRequestBulk.body.length === 0) {
-            return resolve({ids: []});
-          }
+    if (request.input.args.refresh) {
+      esRequest.refresh = true;
+    }
 
-          return this.client.bulk(esRequestBulk)
-            .then(() => this.refreshIndexIfNeeded(esRequestBulk, {ids}))
-            .catch(error => reject(this.esWrapper.formatESError(error)));
-        });
-      });
+    // refresh is needed here to prevent versions conflicts and is acceptable in a delete by query context
+    return this.client.indices.refresh({index: esRequest.index})
+      .then(() => esMethod(esRequest))
+      .then(response => {
+        this.refreshIndexIfNeeded(esRequest, response);
```

Comment:
This will return the result before the resolution of `refreshIndexIfNeeded()` promise.

shoudln't it be rather:
```
then(response => this.refreshIndexIfNeeded(esRequest, {deleted: response[responseMember]}))
```
?

Reply:
oops, you're right. Fixed.